### ISSUE TITLE
Organization of branches, commits, releases and images

### ISSUE THREAD

**Issue by clementgineste** [O-clock-Dev/teleporter-compose#58]
Suggestions pour passer à une approche d'une branche main pour les releases et d'une branche develop pour notre travail de groupe, PR, et aussi la génération d'image en version developpement et non directement sur latest pour ne pas faire passer en "prod" une image pas testée.

* Branches : 
  * `main`, protégée, uniquement sur PR et pas de force push
  * `develop`, protégée, PR aussi comme le fonctionnement actuel sur main. 
* Commit : 
  * [Conventional commit](https://www.conventionalcommits.org/), sera utile par la suite et bien sur nous apporte un standard commun 
* Release : 
  * Automatisée via [release-please](https://github.com/googleapis/release-please) : très sympa, génère tout automatiquement après une PR de release, utilise aussi des labels, génere les releases notes, etc. Se base sur conventional commit
    * Totalement soumis à débat, c'est pas forcément le plus simple à mettre en place bien que dès qu'on maîtrise ça roue tout seul !
* Images Docker : 
  * Build & push bien évidemment automatisés via workflow GHA
  * Différenciation entre `develop` et `latest`
  * Dès que la PR pour merger dans la branche principale et la release automatique alors la `latest` est simplement marquée sur la dernière image `develop` validée

N'hésitez pas à commenter et à ajouter d'autres idées !

<details>
<summary>J'ai demandé à un LLM  (Claude 3 Opus) de me donner d'autres suggestions et voilà sa sortie :</summary>

### Utilisation des GitHub Actions pour l'automatisation

* Tests automatisés : Mettez en place des GitHub Actions pour exécuter automatiquement les tests unitaires et d'intégration à chaque pull request sur les branches develop et main. Cela garantira que seul du code passant tous les tests sera fusionné.
* Vérifications de style de code : Intégrez des actions pour vérifier le respect des conventions de codage à chaque commit, afin de maintenir une haute qualité de code.

### Amélioration de la gestion des issues et pull requests

* Modèles d'issues et PR : Définissez des templates pour les issues et pull requests afin de standardiser leur soumission, avec des champs obligatoires, des checklists et des labels par défaut.
* Bot de gestion : Utilisez un bot comme ceux proposés par Probot pour automatiser certaines tâches répétitives (gestion des conflits, assignation de reviewers, fermeture d'issues inactives...).

### Revue de code et collaboration

* Pair programming : Encouragez la pratique du pair programming, en particulier pour les tâches complexes, en tirant parti d'outils de codage collaboratif intégrés à GitHub.
* Webhooks et notifications : Mettez en place des webhooks pour relier votre dépôt à des outils de communication (Slack, Teams...) et recevoir des notifications en temps réel sur l'activité du dépôt 

### Documentation et wiki

* Documentation des fonctionnalités : Veillez à ce que chaque fonctionnalité clé soit documentée dans le wiki du dépôt, pour aider les nouveaux contributeurs à prendre en main le projet.
* Documentation d'API : Si votre projet expose des API, utilisez des outils comme Swagger ou Postman pour générer une documentation interactive des API, liée à votre dépôt.
</details>

A la fin, il est sympa, il nous souhaite bonne chance. :D

**Comment by clementgoclock** [comment]
WallStreet English incoming :boom: 

---

**Suggestions for moving to a main branch approach for releases and a develop branch for our group work, PRs, and also generating development version images rather than directly on latest to avoid pushing an untested image to "prod".**

* Branches: 
  * **`main`**, protected, only via PR and no force push
  * **`develop`**, protected, PR as well like the current functioning on main
* Commits:
  * **[Conventional commit](https://www.conventionalcommits.org/)**, will be useful later and of course provides us with a common standard
* Release:
  * **Automated via [release-please](https://github.com/googleapis/release-please)**: very nice, generates everything automatically after a release PR, also uses labels, generates release notes, etc. Based on conventional commits
    * Totally open to discussion, it’s not necessarily the easiest to set up but once mastered it runs on its own!
* Docker Images:
  * **Build & push obviously automated via GHA workflow**
  * **Differentiation between `develop` and `latest`**
  * **As soon as the PR to merge into the main branch and the automatic release then `latest` is simply marked on the last validated `develop` image**

Feel free to comment and add other ideas!

<details>
<summary>I asked an LLM (Claude 3 Opus) for other suggestions and here is his output:</summary>

### Using GitHub Actions for Automation

* **Automated tests**: Set up GitHub Actions to automatically run unit and integration tests on each pull request on the develop and main branches. This will ensure that only code passing all tests is merged.
* **Code style checks**: Integrate actions to check code style compliance on each commit, to maintain high code quality.

### Improving Issue and Pull Request Management

* **Issue and PR templates**: Define templates for issues and pull requests to standardize their submission, with required fields, checklists, and default labels.
* **Management bot**: Use a bot like those offered by Probot to automate some repetitive tasks (conflict management, reviewer assignment, closing inactive issues...).

### Code Review and Collaboration

* **Pair programming**: Encourage the practice of pair programming, especially for complex tasks, leveraging collaborative coding tools integrated with GitHub.
* **Webhooks and notifications**: Set up webhooks to connect your repository to communication tools (Slack, Teams...) and receive real-time notifications on repository activity.

### Documentation and Wiki

* **Feature documentation**: Ensure that each key feature is documented in the repository's wiki, to help new contributors get up to speed with the project.
* **API documentation**: If your project exposes APIs, use tools like Swagger or Postman to generate interactive API documentation, linked to your repository.
</details>

At the end, he’s nice, he wishes us good luck. :D

**Comment by clementgoclock** [comment]
Please note that issue https://github.com/O-clock-Dev/teleporter-compose/issues/71 add concept for a building branch (Docker images)

**Comment by clementgoclock** [comment]
As this repo is open source, we can think about using GH's Project function. Only question: can issues be synchronized with two projects (ours, internal, and this one, open)?

**Comment by profy12** [comment]
Yes we can track issues with two projects or more, the projects is at organisation level but can be also public.

Today we are only internals to work on it and we only have privates objectives (lot's of think specifics to Oclock), don't need for public scheduled plan.

**Comment by clementgoclock** [comment]
Just thinking about adding release please to the main branch https://github.com/googleapis/release-please

Proposition:

* We will use develop branch, merge everything in this branch
* When we want to release we'll push to main
* Release please will made the PR for releasing
* RP will do the release
* main will be locked and only use for releasing
